### PR TITLE
WIP Immutable dto support with config

### LINKED ISSUE #30037: [Issue] WIP Immutable dto support with config
This issue is automatically created based on existing pull request: magento/magento2#23265: WIP Immutable dto support with config

---------
Mutable/Immutable DTO support through declarative schema.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by phoenix128**
Mutable/Immutable DTO support through declarative schema.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#30037: WIP Immutable dto support with config

### CODE REVIEW

File: lib/internal/Magento/Framework/Api/DataObjectHelper.php [hunk 1]
```diff
@@ -45,28 +54,46 @@ class DataObjectHelper
...
-        \Magento\Framework\Reflection\TypeProcessor $typeProcessor,
-        \Magento\Framework\Api\ExtensionAttributesFactory $extensionFactory,
-        \Magento\Framework\Api\ExtensionAttribute\JoinProcessorInterface $joinProcessor,
-        MethodsMap $methodsMapProcessor
+        DataObjectProcessor $objectProcessor,
+        TypeProcessor $typeProcessor,
+        ExtensionAttributesFactory $extensionFactory,
+        JoinProcessorInterface $joinProcessor,
+        MethodsMap $methodsMapProcessor,
+        DtoProcessor $dtoProcessor = null,
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Api/DataObjectHelper.php]: New dependencies can be added as required arguments.

File: lib/internal/Magento/Framework/Api/ExtensionAttribute/InjectorProcessorInterface.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Framework\Api\ExtensionAttribute;
+
+/**
+ * Extension attributes injectors processor interface
+ *
+ * @api
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Api/ExtensionAttribute/InjectorProcessorInterface.php]: @api annotation can be added in major release.

File: lib/internal/Magento/Framework/Api/ExtensionAttributesFactory.php [hunk 3]
```diff
@@ -27,48 +38,66 @@ class ExtensionAttributesFactory
...
+     * @param TypeProcessor|null $typeProcessor
      */
-    public function __construct(\Magento\Framework\ObjectManagerInterface $objectManager)
-    {
+    public function __construct(
+        ObjectManagerInterface $objectManager,
+        TypeProcessor $typeProcessor = null
+    ) {
         $this->objectManager = $objectManager;
+        $this->typeProcessor = $typeProcessor ?: ObjectManager::getInstance()
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Api/ExtensionAttributesFactory.php]: The argument can be added as required.

File: lib/internal/Magento/Framework/Dto/Code/Generator/Mutator.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+/**
+ * Class Mutator
+ */
+class Mutator extends EntityAbstract
+{
+    /**
+     * Entity type
+     */
+    const ENTITY_TYPE = 'mutator';
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/Code/Generator/Mutator.php]: Please, add scope visibility for the constant.

File: lib/internal/Magento/Framework/Dto/Code/GetMutatorSourceCode.php [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                        'name' => 'sourceObject',
+                        'description' => $sourceClassName,
+                    ],
+                ]
+            ]
+        ];
+    }
+
+    /**
+     * @param string $resultClassName
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/Code/GetMutatorSourceCode.php:173]: Probably the static test will fail on missed short method's description.

File: lib/internal/Magento/Framework/Dto/Config/Reader.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+declare(strict_types=1);
+
+namespace Magento\Framework\Dto\Config;
+
+use Magento\Framework\Config\Dom;
+use Magento\Framework\Config\FileResolverInterface;
+use Magento\Framework\Config\Reader\Filesystem;
+use Magento\Framework\Config\ValidationStateInterface;
+
+class Reader extends Filesystem
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/Config/Reader.php:18]: Probably the static test will fail on the missed class description.

File: lib/internal/Magento/Framework/Dto/Config/SchemaLocator.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+use Magento\Framework\Config\Dom\UrnResolver;
+use Magento\Framework\Config\SchemaLocatorInterface;
+use Magento\Framework\Exception\NotFoundException;
+
+class SchemaLocator implements SchemaLocatorInterface
+{
+    /**
+     * @var UrnResolver
+     */
+    protected $urnResolver;
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/Config/SchemaLocator.php]: Why the visibility scope is `protected`?

File: lib/internal/Magento/Framework/Dto/DtoProcessor/DtoReflection.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    /**
+     * Return true if a class is extensible through extension attributes
+     *
+     * @param string $className
+     * @return bool
+     * @throws ReflectionException
+     */
+    public function isExtensibleObject(string $className): bool
+    {
+        // TODO: Add a cache layer here
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/DtoProcessor/DtoReflection.php:99]: `@TODO` annotations should be avoided.

File: lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+use Magento\Framework\Reflection\NameFinder;
+use ReflectionException;
+use Zend\Code\Reflection\ClassReflection;
+
+class GetHydrationStrategy
+{
+    /**
+     * Strategy for setter hydration
+     */
+    public const HYDRATOR_STRATEGY_SETTER = 'setter';
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php:22]: Should be these constants as public?

File: lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    /**
+     * Return the strategy for values injection.
+     *
+     *
+     * @param string $className
+     * @param array $data
+     * @return array
+     * @throws ReflectionException
+     */
+    public function execute(string $className, array $data): array
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php:70]: I would like to propose splitting up this method into multiple for better readability.

File: lib/internal/Magento/Framework/Dto/Test/Unit/Code/Generator/DtoGeneratorTest.php [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+class DtoGeneratorTest extends TestCase
+{
+    /**
+     * @var GetDtoSourceCode
+     */
+    private $getDtoSourceCode;
+
+    /**
+     * @var DtoConfig|PHPUnit_Framework_MockObject_MockObject
```
Comment (YevSent) [inline lib/internal/Magento/Framework/Dto/Test/Unit/Code/Generator/DtoGeneratorTest.php]: `PHPUnit\Framework\MockObject\MockObject` should be used instead.

File: dev/tests/integration/testsuite/Magento/Framework/Reflection/Mock/TestExtensionAttributes.php [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Framework\Reflection\Mock;
+
+use Magento\Framework\Api\AbstractSimpleObject;
+use Magento\Framework\Api\ExtensionAttributesInterface;
+
+class TestExtensionAttributes  extends AbstractSimpleObject implements ExtensionAttributesInterface
```
Comment (YevSent) [inline dev/tests/integration/testsuite/Magento/Framework/Reflection/Mock/TestExtensionAttributes.php]: Extra space

File: lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php (hunk 9 above)
Comment (phoenix128) [reply lib/internal/Magento/Framework/Dto/DtoProcessor/GetHydrationStrategy.php:22]: They are required by other classes like `\Magento\Framework\Dto\DtoProcessor`

**Comment by sidolov** [comment]
@magento create issue

**Review by Den4ik — CHANGES_REQUESTED** [review]
@phoenix128 Thanks for work. Is it still actual and WIP?
If yes, please resolve merge conflicts

**Comment by ishakhsuvarov** [comment]
Hi @phoenix128 

I am closing this PR for now as it seems like it is outdated and still needs a lot of work.

As usual, please feel free to reopen if you wish to continue.

Cheers!

**Pull request closed without merging**